### PR TITLE
Yubo fix obvious non editable field in task view

### PR DESCRIPTION
**Pull request by GKD890**
# Description
Frontend_Task143_PR_description

Fixes 16 in **PROJECTS/TASKS/WBS COMPONENT** (medium)

## Main changes explained:

1. Refactor TaskEditModal to some small components to add **Edit** permission. Now in Task View, users who only have **view** permission (e.g. Volunteer user) can only see the contents rather than some inputs .
2. Add a edit permission to **TagsSearch** component to select if to render the input.

## How to test:

check into current branch
do `npm install` and `npm run start:local` to run this PR locally

log as **volunteer** user
go to dashboard→ Click any task in **Task** tab -> click **View** button to see the details of each task.
The fields that **volunteer** can't edit should be obvious.

Also log as **admin** or **manager** , they have either **edit** or **suggest** permission, so the fields can be edited and has `update` button. Also try if **Add Task** button can show the Modal properly.

## Screenshots or videos of changes:
Frontend_PR860_comparing

### Video:

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/57503876/0a7d453a-c5ce-4e4d-961b-c74396f4c8b0

## Note !!! IMPORTANT:
1. Some redundant functions are removed in PR #918 after merged (JUL 8)
2. . If you find such errors like "receive NaN" or "input can't be null", That is due to `calHoursEstimate()` function parse an empty string and it's now fixed (updated: JUN 18).

Screenshot 2023-06-10 at 1 44 01 AM

### CODE REVIEW

**Comment by one-community** [comment]
The before and after images look the same to me here, other than the gray added. Please change everything for people that can't edit this to "hard data". No dropdowns, no fields that look like they can be edited, etc.

**Comment by one-community** [comment]
Changing this PR to "do not review" for now. DM me with an example once this is done.

**Review by 20chen-7 — CHANGES_REQUESTED** [review]
This CSS style looks the same, but the function is intended.
Screenshot 2023-06-01 at 5 12 12 PM

**Review by OAkinbode — COMMENTED** [review]
Hello Yubo. Nice work. The spaces are not editable in volunteer mode as shown below. therefore, the PR functionally works. I think something that others have pointed out is that the UI still "invites" you to try to edit because the fields are still set in text boxes/areas. A suggestion will be to replace those text boxes with perhaps a div without border definitions so that the information appears presented and not editable.  

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/104801652/ef765af1-549e-4075-b9f8-afaf54ec6872

**Review by bienzguoa — DISMISSED** [review]
It works as intended!
For admins, it can be edited and add tasks, for volunteers, it can not be edited or add tasks.

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/92126559/2c704557-4567-4f27-b59f-67c62077a968

**Comment by one-community** [comment]
> Hello Yubo. Nice work. The spaces are not editable in volunteer mode as shown below. therefore, the PR functionally works. I think something that others have pointed out is that the UI still "invites" you to try to edit because the fields are still set in text boxes/areas. A suggestion will be to replace those text boxes with perhaps a div without border definitions so that the information appears presented and not editable.
> 
>  PR.860.working.mov

Yes, please do this

**Comment by GKD890** [comment]
> Hello Yubo. Nice work. The spaces are not editable in volunteer mode as shown below. therefore, the PR functionally works. I think something that others have pointed out is that the UI still "invites" you to try to edit because the fields are still set in text boxes/areas. A suggestion will be to replace those text boxes with perhaps a div without border definitions so that the information appears presented and not editable.
> 
>  PR.860.working.mov

Thank you for suggesting. I was considering to replace that text area, but then I found in **Edit** mode, it's using other library [TinyMCE](https://www.tiny.cloud/docs/tinymce/6/). So the text part is stored in `HTML` format. By only replacing text content may loose format and need to reconstruct whole area. Thus I choose to change the config following their document. But you do remind me that I could still change the background colour. I will try as soon as possible. Thank you.

File: src/components/Projects/WBS/WBSDetail/EditTask/EditTaskModal.jsx [hunk 1]
```diff
@@ -61,13 +61,13 @@ const EditTaskModal = props => {
   const [status, setStatus] = useState('false');
 
   // hour best
-  const [hoursBest, setHoursBest] = useState(0);
```
Comment (GKD890) [inline src/components/Projects/WBS/WBSDetail/EditTask/EditTaskModal.jsx]: Highlight the small changes for reviewing the code. Sorry for inconvenience. 
- Fixed the warnings caused by unnecessary changes.

**Review by cvtqx — DISMISSED** [review]
Works as described. Well done! 

When viewing a task as a volunteer it is in read-only mode and none of the fields are clickable. In admin the input fields are editable.

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/107726913/5a235adc-b48a-4a52-a4e4-5d622ba2b269

**Comment by Jiangwei-shi** [comment]
When I log in as admin, the input fields are editable. It works as indeed.
image

**Review by xaanders — APPROVED** [review]
I have tested the PR. As a volonteer I could see only non-editable information and no button to submit changes:
volunteer
As an admin: 
admin
As a manager: 
manager

Great job! Works as expected.

**Review by scoopstroop92 — COMMENTED** [review]
Hello @GKD890 
The functionality works as intended for a volunteer
However for admin/ manager role I dont think I am able to make changes although the field  looks editable. Kindly check the following video and let me know if i am missing something here 

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/89441435/a07c4dbb-8f76-4b14-8388-af93a3027b3c

File: src/components/Projects/WBS/WBSDetail/EditTask/EditTaskModal.jsx [hunk 2]
```diff
@@ -654,6 +356,552 @@ const EditTaskModal = props => {
...
+      <>
+        <tr>
+          <td scope="col" colSpan="2" className="task-specific-description">
+            Why this Task is Important
+            <Editor
+              init={editorViewInit}
+              name="why-info"
+              className="why-info form-control"
+              value={props.whyInfo}
+            />
```
Comment (freecode23) [inline src/components/Projects/WBS/WBSDetail/EditTask/EditTaskModal.jsx:743]: If you have a hard time configuring the tinyMCE, why not get rid of these `Editor` altogether, since any roles that enters this condition `!props.canEdit` will never need to edit the info anyway?

My thought is just replacing that editor component with:
`<div dangerouslySetInnerHTML={{ __html: props.whyInfo }}></div>`

It will give us this uneditable look shown on the photo below. Then we can just change the css, maybe make it a white background?
Screenshot 2023-07-15 at 9 18 15 PM

The `props.whyInfo` has `<p> </p>` tags around them. That's why I need to use `dangerouslySetInnerHTML` to get rid of those p tags. An alternative to this is just to get rid of `<p> ` using replace() method. Which I think is better if we are use that the whyInfo always only have p tags

**Review by freecode23 — COMMENTED** [review]
1. Tested as 3 roles, owners , admin, and volunteer works as intended
2. Test not hard coded:
Tried change the volunteer permission to be able to edit. Works as intended 
3. Commented on the code to suggest alternative to using tinyMCE just for non-editor roles

File: src/components/Projects/WBS/WBSDetail/EditTask/EditTaskModal.jsx (hunk 2 above)
Comment (GKD890) [reply src/components/Projects/WBS/WBSDetail/EditTask/EditTaskModal.jsx:743]: Just saw the feedback. The methods you suggested is really useful and I didn't know when I submit the PR. As for why not just replacing `<p>` , I was thinking there could be some format left from the person who creates the task. 
However, the tinyMCE will be removed after the new timer deployed, which makes the config for removing tinyMCE no longer important. I will open a new PR with the Prime React Editor ( the new editor) in the future to see if I can fix it well. Thank you for the detailed and clear feedback.

**Review by eduardocodes — COMMENTED** [review]
It works as intended for admins.

When I tried to test for volunteers, I couldn't see tasks from my team I don't know why.

Probably I'm doing something wrong, but at least I managed to test for admins. Good job!


https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/86692035/774babee-8ecb-4594-8142-f579010e6f21

**Review by zuhang2000 — DISMISSED** [review]
Hi I test your PR following your instructions! Everything work as expected. Here are my testing details. 

I first test on Admin account. The task is editable. The mouseover effect is good. The changes are saved once I click update. 

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/123051008/62bd117f-072b-4b31-b633-b245e1aa066d

I also test on Volunteer account. The task is not editable. There is no Update button which is good. 

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/123051008/84517f16-4750-43f6-b27e-842a038295bc

Good job!

**Review by ThapeloMasasa — COMMENTED** [review]
The fields are editable as intended by the PR
Screenshot 2023-08-03 at 1 01 33 PM

**Comment by DavidC0126** [comment]
I tested the PR and it works as intended. When I used a admin account, I was able to edit the task, and when I used a volunteer account I can only view it.
admin_edit

**Review by harshidad — APPROVED** [review]
Working as expected based in user role.

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/109878936/a5e5402a-4d7c-4165-b67d-5aa81ad3a098


https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/109878936/9b8ef89c-e596-463c-962a-87701a382b7b

**Comment by Shubhu0** [comment]
reviewed the pull request and understood all its iterations

**Comment by GaryB93** [comment]
I was able to access the taskEdit modal but unfortunately I was unable to test as admin even logged in with administrator permissions because of the current urgent bug that keeps admin roles from seeing the Other Links tab and pretty much removed my Administrator permissions?

Screenshot 2023-08-28 at 7 23 08 PM

I did however come across some issues building the frontend for this PR and will post them here. I think this could affect performance but not too familiar with it or the app yet. Not sure if this caused the page to be unresponsive or the loading of all tasks as an Admin user.

Screenshot 2023-08-28 at 5 00 15 PM
Screenshot 2023-08-28 at 5 03 13 PM
